Opencast Plugin Manager

lkiesow opened the pull request:
This patch add a new plugin manager, allowing users to easily enable or
disable certain parts of Opencast. This is similar to (and in fact uses)
the Karaf feature configuration with the huge benefit that users to not
have to configure exact versions and keep that updated. Instead they
just need to specify the plugin they want to load.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

lkiesow committed:
Opencast Plugin Manager

This patch add a new plugin manager, allowing users to easily enable or
disable certain parts of Opencast. This is similar to (and in fact uses)
the Karaf feature configuration with the huge benefit that users to not
have to configure exact versions and keep that updated. Instead they
just need to specify the plugin they want to load.

geichelberger commented:
The transcription-service-workflowoperation bundle references the transcription-service-amberscript bundle, which results in a missing requirement if you enable only the google-speech plugin.

lkiesow commented:
Great, sorry. That's definitely a weird dependency. I'll look into it and fix it in some way.

geichelberger commented on modules/plugin-manager/src/main/java/org/opencastproject/plugin/impl/PluginManagerImpl.java:
```diff
@@ -0,0 +1,152 @@
+    // Verbose Karaf logs active?
+    verbose = BooleanUtils.toBoolean(Objects.toString(properties.get(VERBOSE)));
+
+    Thread thread = new Thread(new PluginStarter());
+    thread.start();
```
What happens if the thread is still running and the configuration changes?

geichelberger commented:
This PR #3718 should solve the dependency issue. It would be cleaner to split the workflow operation bundle but having three bundles with just one or two classes seems odd.

lkiesow commented:
Rebased on top of `develop` and fixed all merge conflicts.
Should hopefully now work as expected.

lkiesow commented:
> This PR #3718 should solve the dependency issue. It would be cleaner to split the workflow operation bundle but having three bundles with just one or two classes seems odd.

Seems like that one did not help. Even after the rebase, you get the error. Maybe I just turn this into one transcription service plugin for now…

```
karaf@root()> bundle:list  |grep transc
401 x Active   x  82 x 12.0.0.SNAPSHOT            x Opencast :: transcription-service-api
402 x Active   x  82 x 12.0.0.SNAPSHOT            x Opencast :: transcription-service-google-speech-impl
403 x Active   x  82 x 12.0.0.SNAPSHOT            x Opencast :: transcription-service-persistence
404 x Waiting  x  82 x 12.0.0.SNAPSHOT            x Opencast :: transcription-service-workflowoperation
karaf@root()> bundle:diag 404
Opencast :: transcription-service-workflowoperation (404)
---------------------------------------------------------
Status: Waiting
Declarative Services
org.opencastproject.transcription.workflowoperation.AmberscriptAttachTranscriptionOperationHandler (354)
  missing references: TranscriptionService
org.opencastproject.transcription.workflowoperation.StartTranscriptionOperationHandler (359)
  missing references: TranscriptionService
org.opencastproject.transcription.workflowoperation.AttachTranscriptionOperationHandler (356)
  missing references: TranscriptionService
org.opencastproject.transcription.workflowoperation.AmberscriptStartTranscriptionOperationHandler (355)
  missing references: TranscriptionService

```

geichelberger commented:
Does opencast work? Because the first time, that wiring error caused karaf to fail on startup (If I remember correctly.). The Waiting bundle state is okay because these services have a reference target filter. The Google OSGi Services seem to be active.

geichelberger commented:
A dirty solution would be to make the TranscriptionService references optional.

lkiesow committed:
Simplify Transcription Plugin

For now, this moves all transcription modules into one plugin. We should
split them up later but that needs some restructuring of the workflow
operation module.

lkiesow commented:
For now, I just simplified that by moving the transcription services into one module. We can always split them up more later.

Unfortunately, I found another problem though. Activating and then deactivating the Theodul modules will deactivate all of Opencast's REST endpoints. It is likely a bug of the Theodul modules and not of the plugin manager, but this could still be pretty annoying for users.

My guess would be the cusom OSGi service registration handling in the Theodul plugin manager, but it's too late today to investigate any further.,

geichelberger commented:
I think it has nothing to do with theodul. Its caused by removing the bundle which last created an endpoint. You can reproduce the behavior with just the transcription service. 

It could be caused by the way we createEndpoints and destroy them. If we create an Endpoint we destroy the jax-rs server and recreate it with the new server factory.
```
    if (server != null) {
      logger.debug("Destroying JAX-RS server");
      server.stop();
      server.destroy();
    }

    server = sf.create();
```

After destroying the last servlet, which caused the recreation of the jax-rs server, all other servlets are offline.

https://github.com/opencast/opencast/blob/e57ff7216be8d52e8546ec8cd1b0c09eae835757/modules/kernel/src/main/java/org/opencastproject/kernel/rest/RestPublisher.java#L269-L368

geichelberger commented:
This is a proof of my theory: https://github.com/geichelberger/opencast/commit/595e5021c60fa57702f4a6043ccbdde202f0944d
The proof recreates the JAX-RS server on destroyEndpoint if there are still servlets registered.

With recreating the server, the rest endpoints are still available after deactivating a plugin.

geichelberger commented on docs/guides/admin/docs/modules/plugin-management.md:
```diff
@@ -0,0 +1,32 @@
+- opencast-plugin-transcription-google-speech
+    - Support for the Google cloud transcription service
+- opencast-plugin-transcription-ibm-watson
+    - Support for the IBM cloud transcription service
```
After the latest change, this should only mention the opencast-plugin-transcription-service.

geichelberger commented on docs/guides/admin/docs/modules/plugin-management.md:
```diff
@@ -0,0 +1,32 @@
+Opencast comes with a number of plugins which allow you to turn additional functionality off and on.
+Plugins can be enabled at runtime in `etc/org.opencastproject.plugin.impl.PluginManagerImpl.cfg` by setting the specific
+configuration to `on` or `off`.
+
```
Maybe mention the current limitation that deactivating features at runtime can cause troubles.

lkiesow replied:
done in the config

geichelberger commented on docs/guides/admin/docs/modules/plugin-management.md:
```diff
@@ -0,0 +1,32 @@
+=================
+
+Opencast comes with a number of plugins which allow you to turn additional functionality off and on.
+Plugins can be enabled at runtime in `etc/org.opencastproject.plugin.impl.PluginManagerImpl.cfg` by setting the specific
+configuration to `on` or `off`.
```
I would remove the off, since disabling features does currently causes troubles, and you cannot enable a feature by setting the configuration to off.

lkiesow replied:
Works for most plugins. The only one it does seem to cause trouble is Theodul. Specifically marked this in the config file.

geichelberger commented on assemblies/pom.xml:
```diff
@@ -121,10 +121,16 @@
-              <feature>opencast-brightspace</feature>
+              <feature>opencast-plugin-legacy-annotation</feature>
+              <feature>opencast-plugin-player-theodul</feature>
+              <feature>opencast-plugin-transcription-amberscript</feature>
+              <feature>opencast-plugin-transcription-google-speech</feature>
+              <feature>opencast-plugin-transcription-ibm-watson</feature>
```
opencast-plugin-transcription-services

geichelberger reviewed (changes requested):
LGTM, only some minor changes are required. I would add it to OC 12 even with the problems if you want to deactivate a feature at runtime. The plugin config would not be the only config that can blow up opencast at runtime.

gregorydlogan commented:
I'm going to mark this as on hold, pending you saying it's ready for review again

lkiesow commented:
This is ready for a (hopefully) final review.

geichelberger commented on modules/plugin-manager/src/main/java/org/opencastproject/plugin/impl/PluginManagerImpl.java:
```diff
@@ -0,0 +1,152 @@
+/**
+ * A simple tutorial class to learn about Opencast Services
+ */
+@Component(
+    property = {
+        "service.description=Plugin Manager Service"
+    },
+    immediate = true,
+    service = PluginManagerImpl.class
+)
```
What do you think of using a short and on-point configuration name by setting the configurationPid to
org.opencastproject.plugins

The benefit would be if we change the plugin manager implementation, we could still use the same configuration file.

lkiesow replied:
I don't really care either way. This is consistent with what we use basically everywhere else in Opencast. I also don't expect that we will dramatically change the implementation, but keep the exact same configuration file. Chances are that we need to adjust it, if we change the implementation.

That said, I don't mind changing this, if you want me to. It's a quick and easy change. I'll do whatever it takes to get this merged :grin:  Do you want me to change it?

geichelberger reviewed (changes requested):
LGTM, currently, I only tested the development distribution. I will try the others asap.

geichelberger commented:
I tested the other distributions it is working as expected.

gregorydlogan commented:
@geichelberger says this works, IMO the config file should match the rest of them.  If we're going to start changing naming styles like that we should consider doing *all* of them at the same time, in another PR.  I'm going to merge this so it can finally be in the codebase.
